sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala

```diff
@@ -418,4 +423,10 @@ class JDBCTableCatalog extends TableCatalog
   private def toSQLId(ident: Identifier): String = {
     toSQLId(ident.namespace.toSeq :+ ident.name)
   }
+
+  protected def convertSchemaNameToPattern(schemaName: String): String = {
```

Comment:
is it the same as `V2ExpressionSQLBuilder#escapeSpecialCharsForLikePattern`?

Reply:
Similar to it, but not exactly same, `escapeSpecialCharsForLikePattern` for SqlServer for example can use brackets to escape characters, because default escape character for SqlServer is bracket. `escapeSpecialCharsForLikePattern` have to be paired with `visitStartsWith` and `visitEndsWith`. So some implementation if builder can decide to use different escape characters.
Default implementation of `escapeSpecialCharsForLikePattern` also don't escape /.

Even it is not documented, almost all well-known JDBC drivers (except Oracle) expects pattern with /,%,_ escaping.

Reply:
One advantage of `V2ExpressionSQLBuilder#escapeSpecialCharsForLikePattern` is that each SQL dialect can have its own v2 expression SQL builder, which can override `escapeSpecialCharsForLikePattern` to better support that dialect. Shall we get `JdbcDialect#getJdbcSQLQueryBuilder` and call its `escapeSpecialCharsForLikePattern` method here to do the work?

Reply:
This assumes that the JDBC API of table listing follows the same pattern syntax of LIKE, but I may be wrong. Is there a document for `conn.getMetaData.getTables` that defines the pattern syntax of the schema string?

Reply:
There is no specific documentation about it. Oracle Java JDBC does not describe how pattern looks like, I was already checked that.
Problem with dialect is that / is not escaped, and someone can use different escape character for LIKE operations. Some databases allows that syntax. E.g. SQL Server use brackets for escaping, while `getMetadata`.`getTables` still expects escaping with /. I can add new method in dialect, but I thought there is no reason for that, since this is catalog specific, and because of that, I made this method `protected`, to allow overriding.

Reply:
Do we expect anyone to extend `JDBCTableCatalog`? I don't think so. It's the catalog implementation for all JDBC dialects.

To be safe, we can add a new method in `JdbcDialect` for the escaping in this JDBC API, which is different from the one in `V2ExpressionSQLBuilder`. I hope they are always the same so that we can save an API.

Reply:
> Problem with dialect is that / is not escaped, and someone can use different escape character for LIKE operations.

No, `escapeSpecialCharsForLikePattern` assumes the caller to always use `\` as the escape character.

Reply:
`TableCatalog#listTables` API doc is clear that it should list tables in the given namespace, no pattern syntax should be applied. We shouldn't allow connectors to implement `listTables` differently.

Reply:
The abstraction we need is: the dialect should tell us how to escape special chars in the schema string of `conn.getMetaData.getTables`.

Reply:
note: we don't need this abstraction if there is no difference between dialects, or we can reuse `V2ExpressionSQLBuilder#escapeSpecialCharsForLikePattern`